### PR TITLE
feat: `cabal2nix` integration for IFD-less evalutation

### LINKED ISSUE #144: IFD-less evaluation of local packages
IFD can really slow down the evaluation of the flake, such as to make the launching of `nix develop` noticeably slow -- this is currently the case for https://github.com/nammayatri/nammayatri which has gazillion local packages with lots of dependencies.

The idea here is to do something similar to the `hpack -> cabal` workflow, but for `cabal -> nix`. ie., use `cabal2nix` to generate a `default.nix` in the local package directory ([example](https://github.com/NorfairKing/autodocodec/blob/master/autodocodec/default.nix)), for which we provide a `nix run`'able flake app that the developer will run anytime changing the .cabal file. There will also be a flake check (analogous to #97) to make sure that generated `default.nix` is in sync with the .cabal file (useful in CI).

We could actually name the `default.nix` using something unique to haskell-flake: `haskell-package.nix`.

Then, the following configuration:

```nix
{
  haskellProjects.default = {
    packages = {
      foo.root = ./foo;
    };
}
```

... would look for `haskell-packages.nix` and use that; if not, it looks for the .cabal file falling back to using IFD (`callCabal2nix`).

cc @roberth @locallycompact @NorfairKing for feedback/suggestions/ideas.

### PR DESCRIPTION
**Pull request by rsrohitsingh682**
Addresses #144 just for local packages. IFD will still be used for Hackage packages (`callHackage`) and custom source inputs. 

Example: https://github.com/srid/haskell-template/pull/146

### CODE REVIEW

File: nix/build-haskell-package.nix [hunk 1]
```diff
@@ -22,6 +22,20 @@ let
       name = "${name}";
       inherit src;
     };
+
+  callPackageKeepDeriver = src: args:
```
Comment (srid) [inline nix/build-haskell-package.nix:26]: You want to add a comment explaining why this function is here.

It appears to be a copy of:

https://github.com/NixOS/nixpkgs/blob/d30e9cf3be1cc72f99d5f67772ce961b84126adf/pkgs/development/haskell-modules/make-package-set.nix#L170

Is there no way of directly using it from nixpkgs?

File: nix/build-haskell-package.nix [hunk 2]
```diff
@@ -31,6 +45,19 @@ lib.pipe root
...
     (x: log.traceDebug "${name}.mkNewStorePath ${x.outPath}" x)
 
-    (root: self.callCabal2nix name root { })
-    (x: log.traceDebug "${name}.cabal2nixDeriver ${x.cabal2nixDeriver.outPath}" x)
+     (root:
+        # Check if cached cabal2nix generated nix expression is present,
+        # if present use it with callPackage
+        # to avoid IFD
+        let pkg = 
+          if lib.pathExists (lib.concatStringsSep "/" [ root "default.nix" ])
```
Comment (srid) [inline nix/build-haskell-package.nix:53]: The name of this file should be configurable by the user, much like rust-flake's [`crateNixFile`](https://github.com/juspay/rust-flake/blob/cdcm17a1d009674a9a23c7e4749baf474d2419f04/nix/modules/flake-module.nix#L54).

Also, the default value of this option should not be `default.nix` (people use it for a lot of other things), but rather something ore unique, like `cabal.nix`.

File: nix/build-haskell-package.nix (hunk 1 above)
Comment (rsrohitsingh682) [reply nix/build-haskell-package.nix:26]: Will check that.

File: nix/build-haskell-package.nix (hunk 2 above)
Comment (srid) [reply nix/build-haskell-package.nix:53]: You can open a PR on https://github.com/cachix/git-hooks.nix adding an option that allows the user to override it.

Comment (rsrohitsingh682) [reply nix/build-haskell-package.nix:53]: https://github.com/cachix/git-hooks.nix/pull/531

File: nix/build-haskell-package.nix [hunk 3]
```diff
@@ -22,15 +22,37 @@ let
...
+      else
+        let
+          pkg =
+            (log.traceDebug "${name}.callCabal2nix ${root}")
+              self.callCabal2nix
+              name
+              root
+              { };
+        in
+        (log.traceDebug "${name}.cabal2nixDeriver ${pkg.cabal2nixDeriver.outPath}" pkg)
```
Comment (srid) [inline nix/build-haskell-package.nix:56]: I think should only be logged when `callCabal2nix` is being used.

Comment (rsrohitsingh682) [reply nix/build-haskell-package.nix:56]: Yeah. This is exactly that happens. 
` (log.traceDebug "${name}.cabal2nixDeriver ${pkg.cabal2nixDeriver.outPath}" pkg)` gets logged only when we use `callCabal2nix`.

Logs from `nammayatri`
https://gist.github.com/rsrohitsingh682/243463dbffaeea5fc87f488a997736d1

File: nix/build-haskell-package.nix [hunk 4]
```diff
@@ -22,15 +22,37 @@ let
...
```
Comment (srid) [inline nix/build-haskell-package.nix:56]:
```suggestion
        lib.pipe (self.callPackage path {})
          [ (pkg: log.traceDebug "${name}.callCabal2nix root=${root} deriver=${pkg.cabal2nixDeriver.outPath}")
          ]
```

**Review by srid — COMMENTED** [review]
LGTM. 

As a last thing, could [add a test](https://github.com/srid/haskell-flake/tree/master/test)?

**Review by srid — APPROVED** [review]
LGTM. Just wanting on the other PRs to be merged. And then I'll test it on some of my projects before merging.

**Pull request merged**